src/Microsoft.AspNet.Mvc.Localization/HtmlLocalizer.cs

```diff
@@ -0,0 +1,213 @@
+        public virtual LocalizedString GetString([NotNull] string key, params object[] arguments) =>
+            _localizer.GetString(key, arguments);
+
+        /// <inheritdoc />
+        public virtual IEnumerable<LocalizedString> GetAllStrings(bool includeAncestorCultures) =>
+            _localizer.GetAllStrings(includeAncestorCultures);
+
+        /// <inheritdoc />
+        public virtual LocalizedHtmlString Html([NotNull] string key) => ToHtmlString(_localizer.GetString(key));
+
+        /// <inheritdoc />
+        public virtual LocalizedHtmlString Html([NotNull] string key, params object[] arguments)
+        {
+            var stringValue = _localizer[key].Value;
+
+            return ToHtmlString(new LocalizedString(key, EncodeArguments(stringValue, arguments)));
+        }
+
+        /// <summary>
+        /// Creates a new <see cref="LocalizedHtmlString"/> for a <see cref="LocalizedString"/>.
+        /// </summary>
+        /// <param name="result">The <see cref="LocalizedString"/>.</param>
+        protected virtual LocalizedHtmlString ToHtmlString(LocalizedString result) =>
+            new LocalizedHtmlString(result.Name, result.Value, result.ResourceNotFound);
+
+        /// <summary>
+        /// Encodes the arguments based on the object type.
+        /// </summary>
+        /// <param name="resourceString">The resourceString whose arguments need to be encoded.</param>
+        /// <param name="arguments">The array of objects to encode.</param>
+        /// <returns>The string with encoded arguments.</returns>
+        protected virtual string EncodeArguments([NotNull] string resourceString, [NotNull] object[] arguments)
+        {
+            var position = 0;
+            var length = resourceString.Length;
+            char currentCharacter;
+            StringBuilder tokenBuffer = null;
+            var outputBuffer = new StringBuilder();
+            var isToken = false;
+
+            while (position < length)
+            {
+                currentCharacter = resourceString[position];
+
+                position++;
+                if (currentCharacter == '}')
+                {
+                    if (tokenBuffer == null && !isToken)
+                    {
+                        tokenBuffer = new StringBuilder();
+                        isToken = true;
+                    }
+                    tokenBuffer.Append(currentCharacter);
+
+                    // Treat as escape character for }}
+                    while (position < length && resourceString[position] == '}')
+                    {
+                        tokenBuffer.Append(resourceString[position]);
+                        position++;
+                    }
+
+                    if (tokenBuffer != null && tokenBuffer.Length > 0)
+                    {
+                        var formattedString = string.Format(tokenBuffer.ToString(), arguments);
+                        if (formattedString.StartsWith("{") || formattedString.StartsWith("}"))
```

Comment:
Condition looks fragile but likely covers most mainstream cases. Worth a comment at least, especially because it will fail with some custom format strings e.g. `"{0:'{{characters that should be escaped}}b'###'b'}"`. Note `tokenBuffer != null && tokenBuffer.Length > 0` condition is also incorrect for that custom format string.

However why is this so complex? Would things work correctly to instead copy every other "{" character or "}" character if `!isToken` directly to `outputBuffer` and every "}" character to `tokenBuffer` otherwise?